File: CS50x/sentimental-credit/credit.py
from cs50 import get_int
ccn = 0
while (type(ccn) != int) or ccn == 0:
    ccn = get_int("Number ")
# ccn = str(ccn)
tmp = 0
sumEven = 0
sumOdd = 0
counter = 1
length = len(str(ccn))
number = ccn
if (length < 13 or length > 16):
    print("INVALID")
else:
    while counter <= length:
        # If the digit is in an odd position
        if ((counter % 2) == 1):
            tmp = ccn % 10
            ccn = ccn // 10
            sumOdd += tmp
            #print(f"Odd {tmp}")
            counter += 1

        # If the digit is in an even position
        else:
            tmp = ccn % 10
            ccn = ccn // 10
            #print(f"Even {tmp} {tmp * 2}")
            if tmp < 5:
                sumEven += tmp * 2
            else:
                sumEven += 1
                sumEven += (tmp * 2) % 10
            counter += 1
        # Do it once more to print the last number
    totalSum = sumEven + sumOdd
    if (totalSum % 10 == 0):
        totalSum = True
        # print(totalSum)
        if (length == 15 and (((number // 10000000000000) == 34) or ((number // 10000000000000) == 37))):
            print("AMEX")

        elif (length == 16 and ((number // 100000000000000) in range(51, 56))):
            print("MASTERCARD")

        elif ((length == 16 and ((number // 1000000000000000) == 4)) or (length == 13 and ((number // 1000000000000) == 4))):
            print("VISA")
        else:
            print("INVALID")

    else:
        print("INVALID")

File: CS50x/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

// MASTERCARD: 16-Digit #'s, Start with: 51, 52, 53, 54, or 55
// VISA: 13-16-Digit #'s, Start with: 4
// AMEX: 15-Digit #'s, Start with: 34 or 37

// Luhn's Algorithm:
// 1. Multiply every other digit by 2, starting with the second number to the last
// 2. Add the sum of product's digits. Eg  if the digit is 6 multiply by 2 -> 12 would be 1+2
// 3. Add the sum of the other digits (not multiplied by 2)
// 4. If the total sum ends with a 0, it is valid!



int main(void)
{
    long cardNumber = get_long("Card's number: ");
    long cN = cardNumber;

    //length
    int length = 0;
    do
    {
        cN = cN / 10;
        length ++;
    }
    while (cN != 0);

    //Luhn algorith
    if (length != 13 && length != 15 && length != 16)
    {
        printf("INVALID\n");
        return 0;
    }
    // X is the numbers that are doubled
    int x;
    // Y is the numbers not doubled
    int y;
    int numbersDoubled = 0;
    int numbersNOTDoubled = 0;
    long newNumber = cardNumber;

    for (int i = 0; i < length / 2 + 1; i++)
    {
        y = newNumber % 10;
        //printf("Last number is: %i\n", y);
        numbersNOTDoubled +=  y;
        //printf("numbersNOTDoubled is: %i\n", numbersNOTDoubled);
        newNumber = newNumber / 10;
        //printf("Remaining is: %ld\n\n", newNumber);

        x = newNumber % 10;
        x = x * 2;
        //printf("Last number is: %i\n", x);
        if (x < 10)
        {
            numbersDoubled += x;
        }
        else
        {
            numbersDoubled += 1;
            numbersDoubled += x % 10;
        }

        //printf("numbersDoubled is: %i\n", numbersDoubled);
        newNumber = newNumber / 10;
        //printf("Remaining is: %ld\n\n", newNumber);
    }
    int totalSum = numbersDoubled + numbersNOTDoubled;
    //printf("card is valid\n");
    if (totalSum % 10 == 0 && length == 15)
    {
        int disposableLength = length - 2;
        long divider = pow(10, disposableLength);
        if ((cardNumber / divider) == 37 || (cardNumber / divider) == 34)
        {
            printf("AMEX\n");
            return 0;
        }
    }

    if ((totalSum % 10 == 0 && length == 13) || (totalSum % 10 == 0 && length == 16))
    {
        int disposableLength = length - 1;
        long divider = pow(10, disposableLength);
        if ((cardNumber / divider) == 4)
        {
            printf("VISA\n");
            return 0;
        }
    }

    if (totalSum % 10 == 0 && length == 16)
    {
        int disposableLength = length -2;
        long divider = pow(10, disposableLength);
        if ((cardNumber / divider) == 51 || (cardNumber / divider) == 52 || (cardNumber / divider) == 53 || (cardNumber / divider) == 54 ||
            (cardNumber / divider) == 55)
        {
            printf("MASTERCARD\n");
            return 0;
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}